esphome/components/sen5x/sen5x.cpp

```diff
@@ -0,0 +1,286 @@
+      case FIRMWARE_VERSION_REQUEST_FAILED:
+        ESP_LOGW(TAG, "Unable to request sensor firmware version");
+        break;
+      case FIRMWARE_VERSION_READ_FAILED:
+        ESP_LOGW(TAG, "Unable to read sensor firmware version");
+        break;
+      default:
+        ESP_LOGW(TAG, "Unknown setup error!");
+        break;
+    }
+  }
+  LOG_UPDATE_INTERVAL(this);
+  ESP_LOGCONFIG(TAG, "  Serial Number: '%s'", this->serial_number_);
+  ESP_LOGCONFIG(TAG, "  Firmware version v%0d.%0d", (raw_firmware_version_ >> 8),
+                uint16_t(raw_firmware_version_ & 0xFF));
+  LOG_SENSOR("  ", "PM1.0 Weight Concentration", this->pm_1_0_sensor_);
+  LOG_SENSOR("  ", "PM2.5 Weight Concentration", this->pm_2_5_sensor_);
+  LOG_SENSOR("  ", "PM4 Weight Concentration", this->pm_4_0_sensor_);
+  LOG_SENSOR("  ", "PM10 Weight Concentration", this->pm_10_0_sensor_);
+  LOG_SENSOR("  ", "Ambient Humidity", this->humidity_sensor_);
+  LOG_SENSOR("  ", "Ambient Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "VOC Index", this->voc_sensor_);
+}
+
+void SEN5xComponent::update() {
+  if (!initialized_) {
+    return;
+  }
+  /// Check if warning flag active (sensor reconnected?)
+  if (this->status_has_warning()) {
+    ESP_LOGD(TAG, "Trying to reconnect the sensor...");
+    if (this->write_command_(SEN5x_CMD_RESET)) {
+      ESP_LOGD(TAG, "Sensor has soft-reset successfully. Waiting for reconnection in 1000ms...");
+      this->set_timeout(1000, [this]() {
+        this->start_continuous_measurement_();
+        /// Sensor restarted and reading attempt made next cycle
+        this->status_clear_warning();
+        this->skipped_data_read_cycles_ = 0;
+        ESP_LOGD(TAG, "Sensor reconnected successfully. Resuming continuous measurement!");
+      });
+    } else {
+      ESP_LOGD(TAG, "Sensor soft-reset failed. Is the sensor offline?");
+    }
+    return;
+  }
+  /// Check if measurement is ready before reading the value
+  if (!this->write_command_(SEN5x_CMD_READ_DATA_READY_FLAG)) {
+    this->status_set_warning();
+    return;
+  }
+
+  uint16_t raw_read_status;
+  if (!this->read_data_(&raw_read_status, 1) || raw_read_status == 0x00) {
+    ESP_LOGD(TAG, "Sensor measurement not ready yet.");
+    this->skipped_data_read_cycles_++;
+    /// The following logic is required to address the cases when a sensor is quickly replaced before it's marked
+    /// as failed so that new sensor is eventually forced to be reinitialized for continuous measurement.
+    if (this->skipped_data_read_cycles_ > MAX_SKIPPED_DATA_CYCLES_BEFORE_ERROR) {
+      ESP_LOGD(TAG, "Sensor exceeded max allowed attempts. Sensor communication will be reinitialized.");
+      this->status_set_warning();
+    }
+    return;
+  }
+
+  if (!this->write_command_(SEN5x_CMD_READ_MEASURED_VALUES)) {
+    ESP_LOGW(TAG, "Error reading measurement status!");
+    this->status_set_warning();
+    return;
+  }
+
+  this->set_timeout(50, [this]() {
+    uint16_t raw_data[20];
```

Comment:
Why length 20 here when you only need 8?
```suggestion
    uint16_t raw_data[8];
```

Reply:
Thanks, I missed that